dropshot/src/server.rs

```diff
@@ -850,8 +856,48 @@ async fn http_request_handle<C: ServerContext>(
         request_id: request_id.to_string(),
         log: request_log,
     };
-    let mut response =
-        lookup_result.handler.handle_request(rqctx, request).await?;
+    let handler = lookup_result.handler;
+
+    let mut response = match server.config.default_handler_task_mode {
+        HandlerTaskMode::CancelOnDisconnect => {
+            // For CancelOnDisconnect, we run the request handler directly: if
+            // the client disconnects, we will be cancelled, and therefore this
+            // future will too.
+            handler.handle_request(rqctx, request).await?
+        }
+        HandlerTaskMode::Detached => {
+            // Spawn the handler so if we're cancelled, the handler still runs
+            // to completion.
+            let (tx, rx) = oneshot::channel();
+            let request_log = rqctx.log.clone();
+            tokio::spawn(async move {
+                let request_log = rqctx.log.clone();
+                let result = handler.handle_request(rqctx, request).await;
+
+                // If this send fails, our spawning task has been cancelled in
+                // the `rx.await` below; log such a result.
+                if tx.send(result).is_err() {
+                    warn!(
+                        request_log,
+                        "client disconnected before response returned"
+                    );
+                }
+            });
+
+            // The only way we can fail to receive on `rx` is if `tx` is
+            // dropped before a result is sent, which can only happen if
+            // `handle_request` panics. We will propogate such a panic here,
+            // just as we would have in `CancelOnDisconnect` mode above (where
+            // we call the handler directly).
+            match rx.await {
+                Ok(result) => result?,
+                Err(_) => {
+                    error!(request_log, "handler panicked");
+                    panic!("handler panicked before sending response");
```

Comment:
I'm not sure how to handle this.  It'd be nice to report the error in these cases (or else is it lost completely?).  I'm not clear that we can because we don't have anything usefully printable.

This example is interesting:
https://docs.rs/tokio/latest/tokio/task/struct.JoinError.html#examples-1

Maybe that's the way to go?

Reply:
Nice, that looks great. It's a little more work to get the error out; added in 204c111